File: tests/mathematical/matrix_multiply/matrix_multiply.ts
#!/usr/bin/env node

function createMatrix(rows: number, cols: number): number[][] {
    const matrix: number[][] = [];
    for (let i = 0; i < rows; i++) {
        matrix[i] = [];
        for (let j = 0; j < cols; j++) {
            matrix[i][j] = Math.random() * 100;
        }
    }
    return matrix;
}

function multiplyMatrices(a: number[][], b: number[][]): number[][] {
    const rowsA = a.length;
    const colsA = a[0].length;
    const colsB = b[0].length;
    
    const result: number[][] = [];
    for (let i = 0; i < rowsA; i++) {
        result[i] = new Array(colsB).fill(0);
    }
    
    for (let i = 0; i < rowsA; i++) {
        for (let j = 0; j < colsB; j++) {
            for (let k = 0; k < colsA; k++) {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    
    return result;
}

function main(): void {
    const size = 200; // Matrix size (200x200)
    
    console.log(`Multiplying two ${size}x${size} matrices...`);
    
    // Create matrices
    const createStart = process.hrtime.bigint();
    const matrixA = createMatrix(size, size);
    const matrixB = createMatrix(size, size);
    const createTime = Number(process.hrtime.bigint() - createStart) / 1e9;
    
    // Multiply matrices
    const multiplyStart = process.hrtime.bigint();
    const result = multiplyMatrices(matrixA, matrixB);
    const multiplyTime = Number(process.hrtime.bigint() - multiplyStart) / 1e9;
    
    const totalTime = createTime + multiplyTime;
    
    // Verify result dimensions
    const resultRows = result.length;
    const resultCols = result[0].length;
    
    console.log(`Result: ${resultRows}x${resultCols} matrix`);
    console.log(`Sample result[0][0]: ${result[0][0].toFixed(6)}`);
    console.log("Timing:");
    console.log(`  Matrix creation: ${createTime.toFixed(6)} seconds`);
    console.log(`  Matrix multiplication: ${multiplyTime.toFixed(6)} seconds`);
    console.log(`  Total time: ${totalTime.toFixed(6)} seconds`);
}

main();

File: tests/mathematical/matrix_multiply/matrix_multiply.py
#!/usr/bin/env python3
"""
Matrix multiplication benchmark implementation in Python.
Performs matrix multiplication on randomly generated matrices.
"""

import sys
import time
import random
import numpy as np


def create_matrix(rows, cols):
    """Create a matrix filled with random values using NumPy."""
    return np.random.rand(rows, cols) * 100


def multiply_matrices(a, b):
    """Multiply two matrices using NumPy's optimized multiplication."""
    if a.shape[1] != b.shape[0]:
        raise ValueError("Matrices cannot be multiplied")
    # Use the @ operator for matrix multiplication in NumPy
    return a @ b


def main():
    """Main execution function."""
    size = 200  # Matrix size (200x200)
    
    print(f"Multiplying two {size}x{size} matrices...")
    
    # Create matrices
    create_start = time.perf_counter()
    matrix_a = create_matrix(size, size)
    matrix_b = create_matrix(size, size)
    create_time = time.perf_counter() - create_start
    
    # Multiply matrices
    multiply_start = time.perf_counter()
    result = multiply_matrices(matrix_a, matrix_b)
    multiply_time = time.perf_counter() - multiply_start
    
    total_time = create_time + multiply_time
    
    # Verify result dimensions
    result_rows, result_cols = result.shape
    
    print(f"Result: {result_rows}x{result_cols} matrix")
    print(f"Sample result[0][0]: {result[0][0]:.6f}")
    print(f"Timing:")
    print(f"  Matrix creation: {create_time:.6f} seconds")
    print(f"  Matrix multiplication: {multiply_time:.6f} seconds")
    print(f"  Total time: {total_time:.6f} seconds")


if __name__ == "__main__":
    main()